File: backend/src/test/java/github/lms/lemuel/xr/game/application/ExitSessionUseCaseTest.java
package github.lms.lemuel.xr.game.application;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.Mockito.when;

import github.lms.lemuel.xr.common.AppException;
import github.lms.lemuel.xr.common.ErrorCode;
import github.lms.lemuel.xr.game.adapter.out.persistence.GameSessionJpaEntity;
import github.lms.lemuel.xr.game.adapter.out.persistence.GameSessionRepository;
import java.time.LocalDateTime;
import java.util.Optional;
import java.util.UUID;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class ExitSessionUseCaseTest {

    @Mock GameSessionRepository sessions;
    @InjectMocks ExitSessionUseCase uc;

    private GameSessionJpaEntity live(UUID id) {
        GameSessionJpaEntity e = new GameSessionJpaEntity();
        e.setId(id);
        e.setStartedAt(LocalDateTime.now());
        return e;
    }

    @Test
    void execute_reason_atScene_지정() {
        UUID sid = UUID.randomUUID();
        GameSessionJpaEntity e = live(sid);
        when(sessions.findById(sid)).thenReturn(Optional.of(e));

        var r = uc.execute(sid, "panic", 3);

        assertThat(r.exitedAt()).isNotNull();
        assertThat(e.getAbandonedAt()).isNotNull();
        assertThat(e.getFinalOutcome()).isEqualTo("safe_exit:panic");
        assertThat(e.getSceneCountCompleted()).isEqualTo((short) 3);
    }

    @Test
    void execute_reason_null_이면_user_choice_default() {
        UUID sid = UUID.randomUUID();
        GameSessionJpaEntity e = live(sid);
        when(sessions.findById(sid)).thenReturn(Optional.of(e));

        uc.execute(sid, null, null);

        assertThat(e.getFinalOutcome()).isEqualTo("safe_exit:user_choice");
    }

    @Test
    void execute_세션없음_E_SESSION_NOT_FOUND() {
        UUID sid = UUID.randomUUID();
        when(sessions.findById(sid)).thenReturn(Optional.empty());
        assertThatThrownBy(() -> uc.execute(sid, "x", null))
                .isInstanceOf(AppException.class)
                .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_NOT_FOUND);
    }

    @Test
    void execute_이미완료_E_SESSION_INVALID() {
        UUID sid = UUID.randomUUID();
        GameSessionJpaEntity e = live(sid);
        e.setCompletedAt(LocalDateTime.now());
        when(sessions.findById(sid)).thenReturn(Optional.of(e));
        assertThatThrownBy(() -> uc.execute(sid, "x", null))
                .isInstanceOf(AppException.class)
                .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_INVALID);
    }

    @Test
    void execute_이미abandoned_E_SESSION_INVALID() {
        UUID sid = UUID.randomUUID();
        GameSessionJpaEntity e = live(sid);
        e.setAbandonedAt(LocalDateTime.now());
        when(sessions.findById(sid)).thenReturn(Optional.of(e));
        assertThatThrownBy(() -> uc.execute(sid, "x", null))
                .isInstanceOf(AppException.class)
                .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_INVALID);
    }
}

File: backend/src/test/kotlin/github/lms/lemuel/xr/game/application/ExitSessionUseCaseTest.kt
package github.lms.lemuel.xr.game.application

import github.lms.lemuel.xr.common.AppException
import github.lms.lemuel.xr.common.ErrorCode
import github.lms.lemuel.xr.game.application.port.out.GameSessionPort
import github.lms.lemuel.xr.game.domain.GameSession
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.time.LocalDateTime
import java.util.Optional
import java.util.UUID

class ExitSessionUseCaseTest {

    private val sessions: GameSessionPort = mock()
    private val owner: UUID = UUID.randomUUID()
    private val uc = ExitSessionUseCase(sessions)

    private fun live(id: UUID): GameSession =
        GameSession.reconstitute(
            id, owner, null, null, null,
            LocalDateTime.now(), null, null, null, null, null, null,
            0.toShort(), null, null, null, null,
        )

    private fun completed(id: UUID): GameSession =
        GameSession.reconstitute(
            id, owner, null, null, null,
            LocalDateTime.now(), LocalDateTime.now(), null, null, null, null, null,
            0.toShort(), null, null, null, null,
        )

    private fun abandoned(id: UUID): GameSession =
        GameSession.reconstitute(
            id, owner, null, null, null,
            LocalDateTime.now(), null, LocalDateTime.now(), null, null, null, null,
            0.toShort(), null, null, null, null,
        )

    @Test
    fun `execute reason atScene 지정`() {
        val sid = UUID.randomUUID()
        val e = live(sid)
        whenever(sessions.findById(sid)).thenReturn(Optional.of(e))

        val r = uc.execute(owner, sid, "panic", 3)

        assertThat(r.exitedAt).isNotNull()
        assertThat(e.abandonedAt).isNotNull()
        assertThat(e.finalOutcome).isEqualTo("safe_exit:panic")
        assertThat(e.sceneCountCompleted).isEqualTo(3.toShort())
    }

    @Test
    fun `execute reason null 이면 user_choice default`() {
        val sid = UUID.randomUUID()
        val e = live(sid)
        whenever(sessions.findById(sid)).thenReturn(Optional.of(e))

        uc.execute(owner, sid, null, null)

        assertThat(e.finalOutcome).isEqualTo("safe_exit:user_choice")
    }

    @Test
    fun `execute 세션없음 E_SESSION_NOT_FOUND`() {
        val sid = UUID.randomUUID()
        whenever(sessions.findById(sid)).thenReturn(Optional.empty())
        assertThatThrownBy { uc.execute(owner, sid, "x", null) }
            .isInstanceOf(AppException::class.java)
            .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_NOT_FOUND)
    }

    @Test
    fun `execute 이미완료 E_SESSION_INVALID`() {
        val sid = UUID.randomUUID()
        whenever(sessions.findById(sid)).thenReturn(Optional.of(completed(sid)))
        assertThatThrownBy { uc.execute(owner, sid, "x", null) }
            .isInstanceOf(AppException::class.java)
            .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_INVALID)
    }

    @Test
    fun `execute 이미abandoned E_SESSION_INVALID`() {
        val sid = UUID.randomUUID()
        whenever(sessions.findById(sid)).thenReturn(Optional.of(abandoned(sid)))
        assertThatThrownBy { uc.execute(owner, sid, "x", null) }
            .isInstanceOf(AppException::class.java)
            .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_INVALID)
    }

    // --- IDOR 회귀 ---

    @Test
    fun `남의 세션은 종료할 수 없다 - 존재를 숨기려 404`() {
        val sid = UUID.randomUUID()
        val e = live(sid)
        whenever(sessions.findById(sid)).thenReturn(Optional.of(e))

        assertThatThrownBy { uc.execute(UUID.randomUUID(), sid, "panic", 3) }
            .isInstanceOf(AppException::class.java)
            .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_NOT_FOUND)

        // 거부만으로는 부족하다 — 상태가 변하지 않았음을 확인한다.
        assertThat(e.abandonedAt).isNull()
        assertThat(e.finalOutcome).isNull()
    }

    @Test
    fun `userId 없는 레거시 세션은 아무도 종료할 수 없다`() {
        val sid = UUID.randomUUID()
        val legacy = GameSession.reconstitute(
            sid, null, null, null, null,
            LocalDateTime.now(), null, null, null, null, null, null,
            0.toShort(), null, null, null, null,
        )
        whenever(sessions.findById(sid)).thenReturn(Optional.of(legacy))

        assertThatThrownBy { uc.execute(owner, sid, "x", null) }
            .isInstanceOf(AppException::class.java)
            .hasFieldOrPropertyWithValue("code", ErrorCode.E_SESSION_NOT_FOUND)
    }
}
